src/CourseProgress.php

```diff
@@ -0,0 +1,147 @@
+<?php
+
+namespace Drupal\anu_lms;
+
+use Drupal\node\NodeInterface;
+use Drupal\Core\Url;
+
+/**
+ * Group together progress-related methods.
+ */
+class CourseProgress {
+
+  /**
+   * The course service.
+   *
+   * @var \Drupal\anu_lms\Course
+   */
+  protected $course;
+
+  /**
+   * The lesson service.
+   *
+   * @var \Drupal\anu_lms\Lesson
+   */
+  protected $lesson;
+
+  /**
+   * The courses page service.
+   *
+   * @var \Drupal\anu_lms\CoursesPage
+   */
+  protected $coursesPage;
+
+  /**
+   * Constructs service.
+   *
+   * @param \Drupal\anu_lms\Course $course
+   *   The course service.
+   * @param \Drupal\anu_lms\Lesson $lesson
+   *   The lesson service.
+   * @param \Drupal\anu_lms\CoursesPage $coursesPage
+   *   The courses page service.
+   */
+  public function __construct(Course $course, Lesson $lesson, CoursesPage $coursesPage) {
+    $this->course = $course;
+    $this->lesson = $lesson;
+    $this->coursesPage = $coursesPage;
+  }
+
+  /**
+   * Returns progress in a course.
+   *
+   * @param \Drupal\node\NodeInterface $course
+   *   Course object.
+   *
+   * @return float
+   *   The progress as a percentage like 33.34.
+   */
+  public function getCourseProgress(NodeInterface $course) {
+    $modules = $course->get('field_course_module')->referencedEntities();
+    $totalLessons = 0;
+    $completedLessons = 0;
+    foreach ($modules as $module) {
+
+      /** @var \Drupal\node\NodeInterface[] $lessons */
+      $lessons = $module->field_module_lessons->referencedEntities();
+      foreach ($lessons as $lesson) {
+        if ($lesson->access('view')) {
+          $totalLessons++;
+          if ($this->lesson->isCompleted($lesson)) {
+            $completedLessons++;
+          }
+        }
+      }
+      if (!$module->field_module_assessment) {
+        continue;
+      }
+      /** @var \Drupal\node\NodeInterface[] $quizzes */
+      $quizzes = $module->field_module_assessment->referencedEntities();
+      foreach ($quizzes as $quiz) {
+        if ($quiz->access('view')) {
+          $totalLessons++;
+          if ($this->lesson->isCompleted($quiz)) {
+            $completedLessons++;
+          }
+        }
+      }
```

Comment:
i would extract this into a function called `count_quizzes` or something similar

Reply:
Added as a method in the course service